Refactor enirouting package to reduce code interdependence and add test coverage

christarazi opened the pull request:
See commit messages.

Follow-up refactor from previous PR: https://github.com/cilium/cilium/pull/11073

christarazi commented on pkg/aws/eni/routing/info.go:
```diff
@@ -0,0 +1,91 @@
+// NewRoutingInfo creates a new RoutingInfo struct from a source which adheres
+// to the getter interface. The data from the source is then parsed and
+// validated.
+func NewRoutingInfo(getter getter) (*RoutingInfo, error) {
```
@joestringer I've tried out your suggestion from: https://github.com/cilium/cilium/pull/11073#discussion_r411797976

After implementing it, I'm not sure what benefits we get from abstracting away the source of the info via the `getter` interface. I think we'd get the same functionality and non-code-interdependence from just having the `NewRoutingInfo` take the raw values as parameters. 

Is there something I'm missing that the interface would buy us?

joestringer replied:
Thanks for trying it out, good question :)

Looking closer, in this case the ability to fetch the routing info, ie the implementation of `getter` is trivial; it's just taking the API response structure and returning the values inside. So in this current implementation I'm inclined to agree that the benefit is not obvious. Right now, it allows `pkg/aws/eni/routing` to avoid importing the API structures, but I don't think that's a big deal; we're already assuming a model around the cloud IPAM API responses in all of the ENI code structure anyway so the abstraction doesn't help.

If the full implementation of that getter was more complex (For example actually involved pulling values from an API) then I think it would make a bigger difference.

Looks like we got more value out of the `routingConfigurer` above, which allowed us to remove ENI references from that package (module the variable names, which should probably be simplified).

joestringer replied:
Oh, perhaps the unsaid piece above: With this style of implementation, you can create your own dummy `getter` and actually unit-test the `parse()` function below trivially. In this specific case, if we remove it, then you just have to instantiate IPAM API response structures in the tests which is probably still straightforward.

If the `getter` was more complex then just to unit test this functionality it would be more work.

christarazi replied:
Right. The retrieval of these ENI routing values is trivial and will probably continue to be, since it's inherently a cloud API interaction. So our `getter` isn't helping by providing any useful abstraction.

I'm going to take out the `getter` and replace the constructor with raw string values as input instead. This will still allow us to unit-test `parse` easily since it will just take those values as well.

christarazi replied:
> (module the variable names, which should probably be simplified).

Which variable names need to be simplified?

joestringer replied:
Oh, I mean `eniHealthRouting` -> `routeConfiguration`. This assumes an eni implementation but it doesn't make sense to assume that when a generic routeConfigurer is handed down.

christarazi committed:
Refactor enirouting pkg to use interfaces

This commit structures the code in such a way that allows the previously
duplicated parsing/validation logic to be consolidated into one place. It also
reduces the interdependencies between packages and allows for a dummy
implementation to be used for testing purposes.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
Add unit-tests for enirouting package

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
Return error when deleting ENI rules fails

Previously, we relied on a log message to indicate whether a deletion of
a rule failed. Relying on the log message can be potentially misleading
as endpoint deletion is best-effort and errors are ignored in the end.
The user may see log message indicating a rule was deleted, but later in
the endpoint deletion it would show as an error.

This commit changes the function which deletes rules to surface the
error. This change removes the misleading log messages and allows us to
assert against this in tests.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
